File: app/src/main/java/ru/p3tr0vich/fuel/SyncAccount.java
package ru.p3tr0vich.fuel;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.content.ContentResolver;
import android.content.Context;
import android.text.TextUtils;

class SyncAccount {

    private static final String TAG = "SyncAccount";

    private static final String YANDEX_DISK_TOKEN = "yandex disk token";

    private final AccountManager mAccountManager;

    private final String mAuthority;
    private final String mAccountName;
    private final String mAccountType;

    private final Account mAccount;

    SyncAccount(Context context) {
        mAccountManager = (AccountManager) context.getSystemService(Context.ACCOUNT_SERVICE);

        mAuthority = context.getString(R.string.sync_authority);
        mAccountName = context.getString(R.string.sync_account_name);
        mAccountType = context.getString(R.string.sync_account_type);

        mAccount = createAccount();
    }

    private Account createAccount() {
        Account account;

        Account accounts[] = mAccountManager.getAccountsByType(getAccountType());

        if (accounts.length > 0) {
            account = accounts[0];

//            Functions.logD("SyncAccount -- getAccount: accounts[0] == " + accounts[0].toString());
        } else {
            account = new Account(getAccountName(), getAccountType());

            if (mAccountManager.addAccountExplicitly(account, null, null)) {
                setIsSyncable(account, PreferenceManagerFuel.isSyncEnabled());

                UtilsLog.d(TAG, "createAccount", "addAccountExplicitly == true");
            } else {
                UtilsLog.d(TAG, "createAccount", "addAccountExplicitly == false");
            }
        }

        return account;
    }

    public String getAuthority() {
        return mAuthority;
    }

    private String getAccountName() {
        return mAccountName;
    }

    private String getAccountType() {
        return mAccountType;
    }

    public Account getAccount() {
        return mAccount;
    }

    public boolean isSyncActive() {
        return ContentResolver.isSyncActive(getAccount(), getAuthority());
    }

    private void setIsSyncable(final Account account, final boolean syncable) {
        ContentResolver.setIsSyncable(account, getAuthority(), syncable ? 1 : 0);
    }

    public void setIsSyncable(final boolean syncable) {
        setIsSyncable(getAccount(), syncable);
    }

    @SuppressWarnings("SameParameterValue")
    private String getUserData(final String key) {
        return mAccountManager.getUserData(getAccount(), key);
    }

    @SuppressWarnings("SameParameterValue")
    private void setUserData(final String key, final String value) {
        mAccountManager.setUserData(getAccount(), key, value);
    }

    public String getYandexDiskToken() {
        return getUserData(YANDEX_DISK_TOKEN);
    }

    public boolean isYandexDiskTokenEmpty() {
        return TextUtils.isEmpty(getYandexDiskToken());
    }

    public void setYandexDiskToken(final String token) {
        setUserData(YANDEX_DISK_TOKEN, token);
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/sync/SyncAccount.kt
package ru.p3tr0vich.fuel.sync

import android.accounts.Account
import android.content.ContentResolver
import android.content.Context
import ru.p3tr0vich.fuel.R
import ru.p3tr0vich.fuel.helpers.PreferencesHelper
import ru.p3tr0vich.fuel.helpers.SystemServicesHelper
import ru.p3tr0vich.fuel.utils.UtilsLog

class SyncAccount(context: Context) {

    private val accountManager = SystemServicesHelper.getAccountManager(context)!!

    val authority: String = context.getString(R.string.sync_authority)

    private val accountName = context.getString(R.string.sync_account_name)
    private val accountType = context.getString(R.string.sync_account_type)

    val account = createAccount(context)

    val isSyncActive: Boolean
        get() = ContentResolver.isSyncActive(account, authority)

    var yandexDiskToken: String?
        get() = accountManager.getUserData(account, YANDEX_DISK_TOKEN)
        set(value) = accountManager.setUserData(account, YANDEX_DISK_TOKEN, value)

    private fun createAccount(context: Context): Account {
        val account: Account

        val accounts = accountManager.getAccountsByType(accountType)

        if (accounts.isNotEmpty()) {
            account = accounts[0]
        } else {
            account = Account(accountName, accountType)

            if (accountManager.addAccountExplicitly(account, null, null)) {
                setIsSyncable(account, PreferencesHelper.getInstance(context).isSyncEnabled)

                UtilsLog.d(TAG, "createAccount", "addAccountExplicitly == true")
            } else {
                UtilsLog.d(TAG, "createAccount", "addAccountExplicitly == false")
            }
        }

        return account
    }

    private fun setIsSyncable(account: Account, syncable: Boolean) {
        ContentResolver.setIsSyncable(account, authority, if (syncable) 1 else 0)
    }

    fun setIsSyncable(syncable: Boolean) {
        setIsSyncable(account, syncable)
    }

    companion object {
        private const val TAG = "SyncAccount"

        private const val YANDEX_DISK_TOKEN = "yandex disk token"
    }
}
